compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -2319,7 +2319,41 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                         let cast_ty_from = CastTy::from_ty(ty_from);
                         let cast_ty_to = CastTy::from_ty(*ty);
                         match (cast_ty_from, cast_ty_to) {
-                            (Some(CastTy::Ptr(_)), Some(CastTy::Ptr(_))) => (),
+                            (Some(CastTy::Ptr(src)), Some(CastTy::Ptr(dst))) => {
+                                let src_tail = tcx.struct_tail_without_normalization(src.ty);
+                                let dst_tail = tcx.struct_tail_without_normalization(dst.ty);
+
+                                if let ty::Dynamic(..) = src_tail.kind()
+                                    && let ty::Dynamic(dst_tty, ..) = dst_tail.kind()
+                                    && dst_tty.principal().is_some()
```

Comment:
Add a comment that this check exists for ensuring trait object -> trait object casts always have a valid vtable (which is irrelevant if there is no principal, aka only marker traits with no methods).

Reply:
I think _marker_ traits count as principal, it's auto traits that don't. Will add a comment though.

Reply:
Yeah, oli's comment but `s/marker/auto/`